File: ride-sharing-backend/src/main/java/hu/ridesharing/service/RatingService.java
package hu.ridesharing.service;

import hu.ridesharing.dto.RatingDTO;
import hu.ridesharing.dto.response.outgoing.ResponseStatus;
import hu.ridesharing.entity.*;
import hu.ridesharing.exception.RatingException;
import hu.ridesharing.repository.RatingRepository;
import io.micrometer.common.util.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

@Service
public class RatingService {

    private final RatingRepository ratingRepository;

    private final JourneyService journeyService;

    @Autowired
    public RatingService(RatingRepository ratingRepository, JourneyService journeyService) {
        this.ratingRepository = ratingRepository;
        this.journeyService = journeyService;
    }

    public long getMyRatingCount(String username) {
        User user = new User();
        user.setUsername(username);
        return this.ratingRepository.countByRated(user) + this.ratingRepository.countByRater(user);
    }

    public Page<RatingDTO> getReceivedAsPassenger(int page, String username) {
        User passenger = new User();
        passenger.setUsername(username);
        return ratingRepository.findByPassengerRated(passenger, PageRequest.of(0, (page + 1) * 5))
                .map(this::mapToDTO);
    }

    public Page<RatingDTO> getReceivedAsDriver(int page, String username) {
        User driver = new User();
        driver.setUsername(username);
        return ratingRepository.findByDriverRated(driver, PageRequest.of(0, (page + 1) * 5))
                .map(this::mapToDTO);
    }

    public Page<RatingDTO> getGivenAsDriver(int page, String username) {
        User driver = new User();
        driver.setUsername(username);
        return ratingRepository.findByDriverRater(driver, PageRequest.of(0, (page + 1) * 5))
                .map(this::mapToDTO);
    }

    public Page<RatingDTO> getGivenAsPassenger(int page, String username) {
        User passenger = new User();
        passenger.setUsername(username);
        return ratingRepository.findByPassengerRater(passenger, PageRequest.of(0, (page + 1) * 5))
                .map(this::mapToDTO);
    }

    public ResponseStatus rateMyPassenger(Long driveId, Rating rating, String username) {
        if (StringUtils.isBlank(rating.getComment())) {
            throw new RatingException("Comment must not be empty");
        }

        Journey journey = journeyService.checkMyJourney(username, driveId);

        User driver = journey.getDriver();

        RatingId id = new RatingId();
        id.setRated(rating.getRated().getUsername());
        id.setRater(driver.getUsername());
        id.setJourney(journey.getId());
        id.setType(RatingType.DRIVER_TO_PASSENGER);

        if (ratingRepository.existsById(id)) {
            throw new RatingException("You have already rated this passenger.");
        }

        rating.setJourney(journey);
        rating.setRater(driver);
        rating.setType(RatingType.DRIVER_TO_PASSENGER);
        ratingRepository.save(rating);
        return new ResponseStatus(true);
    }

    public ResponseStatus rateMyDriver(Long rideId, Rating rating, String username) {
        if (StringUtils.isBlank(rating.getComment())) {
            throw new RatingException("Comment must not be empty");
        }

        Journey journey = journeyService.checkMyJourney(username, rideId);

        User passenger = new User();
        passenger.setUsername(username);

        RatingId ratingId = new RatingId();
        ratingId.setRated(journey.getDriver().getUsername());
        ratingId.setRater(passenger.getUsername());
        ratingId.setJourney(journey.getId());
        ratingId.setType(RatingType.PASSENGER_TO_DRIVER);

        if (ratingRepository.existsById(ratingId)) {
            throw new RatingException("You have already rated this driver.");
        }

        rating.setJourney(journey);
        rating.setRater(passenger);
        rating.setType(RatingType.PASSENGER_TO_DRIVER);
        ratingRepository.save(rating);
        return new ResponseStatus(true);
    }

    private RatingDTO mapToDTO(Rating rating) {
        RatingDTO dto = new RatingDTO();
        dto.setValue(rating.getValue());
        dto.setComment(rating.getComment());
        dto.setRated(rating.getRated());
        dto.setRater(rating.getRater());
        dto.setType(rating.getType());
        return dto;
    }
}

File: ride-sharing-frontend/src/app/services/rating-service.ts
import {inject, Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {RatingResponseModel} from '../model/rating-response-model';

@Injectable({
  providedIn: 'root'
})
export class RatingService {

  private readonly url = "/api/rating";

  private readonly httpClient = inject(HttpClient);

  public getReceivedAsPassenger(page: number) {
    return this.httpClient.get<RatingResponseModel>(`${this.url}/received/passenger?page=${page}`);
  }

  public getReceivedAsDriver(page: number) {
    return this.httpClient.get<RatingResponseModel>(`${this.url}/received/driver?page=${page}`);
  }

  public getGivenAsPassenger(page: number) {
    return this.httpClient.get<RatingResponseModel>(`${this.url}/given/passenger?page=${page}`);
  }

  public getGivenAsDriver(page: number) {
    return this.httpClient.get<RatingResponseModel>(`${this.url}/given/driver?page=${page}`);
  }

  public addRatingAsDriver(driveId: number, username: string, rating: number, comment: string) {
    const model = {rated: {username: username}, value: rating, comment: comment};
    return this.httpClient.post<ResponseStatus>(`${this.url}/add/as-driver/${driveId}`, model);
  }

  public addRatingAsPassenger(rideId: number, username: string, rating: number, comment: string) {
    const model = {rated: {username: username}, value: rating, comment: comment};
    return this.httpClient.post<ResponseStatus>(`${this.url}/add/as-passenger/${rideId}`, model);
  }

  public getMyRatingCount() {
    return this.httpClient.get<number>(`${this.url}/my-count`);
  }
}
